### PR TITLE
use __make_integer_seq intrinsic on compilers that support it

### PR DESCRIPTION
**Pull request by ericniebler**
Clang and MSVC have compiler intrinsics for building variadic integer sequences. Use them! Also, tweak `FATAL_BUILD_STR_IMPL` to instantiate fewer templates. This is good for a roughly 2% compile time bump on projects that heavily use `FATAL_STR`.

### CODE REVIEW

File: fatal/type/sequence.h [hunk 1]
```diff
@@ -1004,32 +1021,30 @@ struct polynomial<T, constant_sequence<T>, Variable, VariableAccumulator, Accumu
 
 template <typename T, std::size_t Size>
 constexpr std::size_t size(T const (&)[Size]) {
-  static_assert(
-    Size > 0,
-    "expecting a string containing at least the null terminator"
-  );
```
Comment (ericniebler) [inline fatal/type/sequence.h]: Arrays must have at least 1 element. This check is superfluous.

Comment (juchem) [inline fatal/type/sequence.h]: apparently not: http://ideone.com/SgNS8n

**Comment by juchem** [comment]
This pull request fails contbuild. Please, check what's going on.

**Comment by ericniebler** [comment]
Looks unrelated to my change:

```
$ curl -sSL "http://llvm.org/apt/llvm-snapshot.gpg.key" | sudo -E apt-key add -
gpg: no valid OpenPGP data found.
The command "curl -sSL "http://llvm.org/apt/llvm-snapshot.gpg.key" | sudo -E apt-key add -" failed and exited with 2 during .
```

Something wrong with the Travis job?

File: fatal/type/sequence.h (hunk 1 above)
Comment (ericniebler) [inline fatal/type/sequence.h]:
```sh
eniebler-mbp:~ eniebler$ ~/llvm-install/bin/clang++ -pedantic -Wall -Wextra -std=gnu++11 -c test.cpp
test.cpp:2:14: warning: zero size arrays are an extension [-Wzero-length-array]
  int rg[] = {};
             ^
1 warning generated.
```

It's not legal C++. Looks like compilers implement this as an extension. I agree the assert should stay.

**Comment by ericniebler** [comment]
Ah. If you go to http://llvm.org/apt/ it says:

```
APT mirror was temporary switched off due to excess load. We are working on bringing it back. Stay tuned!
```

**Comment by ericniebler** [comment]
Doesn't look like llvm.org/apt will be coming back online any time soon: https://stackoverflow.com/questions/37558151/travis-ci-build-failing

All your CI builds will fail. I don't think that it should hold up this PR.

File: fatal/type/sequence.h [hunk 2]
```diff
@@ -1012,24 +1029,26 @@ constexpr std::size_t size(T const (&)[Size]) {
...
 
 #define FATAL_BUILD_STR_IMPL(Id, String, Class, Indexes) \
-  template <::std::size_t... Indexes> \
   struct Class { \
+    using value_type = typename ::std::decay<decltype(*(String))>::type; \
+    \
+  private: \
+    template <::std::size_t... Indexes> \
+    static ::fatal::constant_sequence<value_type, (String)[Indexes]...> \
+      make_string_(::fatal::size_sequence<Indexes...> *); \
```
Comment (juchem) [inline fatal/type/sequence.h]: Fatal doesn't append `_` for methods. `s/make_string_/make_string/`

**Comment by juchem** [comment]
I'll run contbuild locally.

**Comment by ericniebler** [comment]
👍

**Pull request merged**